src/raft/recv_install_snapshot.c

```diff
@@ -478,32 +491,116 @@ static void follower_sent_cb(struct sender *s, int rc)
 	struct rpc *rpc = CONTAINER_OF(s, struct rpc, sender);
 	struct follower *follower = CONTAINER_OF(rpc, struct follower, rpc);
 
+	PRE(follower->ops->is_main_thread());
 	if (UNLIKELY(rc != 0)) {
 		sm_move(&rpc->sm, RPC_ERROR);
 		return;
 	}
-
 	follower_tick(follower, M_MSG_SENT);
 }
 
 static bool is_a_trigger_leader(const struct leader *leader, const struct raft_message *incoming)
```

Comment:
Does it seem less noisy? What do you think about such a style?

```
static bool is_a_trigger_leader(const struct leader *leader, const struct raft_message *incoming)
{
        int state = sm_state(&leader->sm);

	return
		ergo(incoming == M_WORK_DONE,
		     IN(state, (LS_HT_WAIT, LS_PAGE_SENT,
				LS_PERSISTED_SIG_PART, LS_PAGE_READ))) &&
		ergo(incoming == M_MSG_SENT,
		     sm_state(&leader->rpc.sm) != RPC_FILLED &&
		     IN(state, (LS_PAGE_READ, LS_SNAP_DONE, LS_F_NEEDS_SNAP,
				LS_REQ_SIG_LOOP, LS_CHECK_F_HAS_SIGS))) &&
		ergo(incoming == M_TIMEOUT,
		     IN(state, (LS_PAGE_READ, LS_SNAP_DONE, LS_F_NEEDS_SNAP,
				LS_REQ_SIG_LOOP, LS_CHECK_F_HAS_SIGS,
				LS_WAIT_SIGS))) &&
		ergo(state == LS_F_ONLINE,
		     incoming->type == RAFT_IO_APPEND_ENTRIES_RESULT) &&
		ergo(sm_state(&leader->rpc.sm) != RPC_SENT, false) &&
		ergo(state == LS_PAGE_READ,
		     IN(incoming->type, (RAFT_IO_INSTALL_SNAPSHOT_CP_RESULT
					 RAFT_IO_INSTALL_SNAPSHOT_MV_RESULT))) &&
		ergo(IN(state, (LS_SNAP_DONE, LS_F_NEEDS_SNAP)),
		     incoming->type == RAFT_IO_INSTALL_SNAPSHOT_RESULT) &&
		ergo(IN(state, (LS_REQ_SIG_LOOP, LS_CHECK_F_HAS_SIGS)),
		     incoming->type == RAFT_IO_SIGNATURE_RESULT);
}
```
cc @cole-miller

Reply:
If we are going for this style I would prefer to have two distinct  blocks because in the first three `ergos` the pointer is not valid `M_WORK_DONE`, `M_MSG_SENT` and `M_TIMEOUT` are equal to `1`, `2` and `3`, which means that doing `incoming->type` will crash the program.

My fear is that when we come back and do changes in the future we miss that bit about the ordering because it is not explicit.